### PR TITLE
Detect and use ccache

### PR DESCRIPTION
**Pull request by Hi-Angel**
One feature that meson does but CMake doesn't is [automatically detecting and using ccache](https://mesonbuild.com/Feature-autodetection.html). It's a great feature, and greatly speeds up recompilations that often caused by common development activity, such as rebasing branches, checking out stuff, moving commits around, etc. Let's fix this by wiring up ccache autodetection *(based on [this stackoverflow answer](https://stackoverflow.com/a/24305849/2388257#how-to-use-ccache-with-cmake))*.

### CODE REVIEW

**Comment by Alayan-stk-2** [comment]
@deveee @Benau Does one of you intend to review this ?

**Comment by deveee** [comment]
Tbh. I always use a script with something like
```
		export CC="ccache gcc"
		export CXX="ccache g++"
```
and I prefer that using ccache is user's decision rather than do it automatically. But anyway no strong opinion...

**Comment by Hi-Angel** [comment]
@deveee but for this people need to remember that the particular build system supports variables, and to know about existence of ccache *(one could have occasionally installed it, e.g. while experimenting with stuff)*. Someone may have STK among dozens of other FOSS projects they sometimes work on *(or just compile, e.g. if a user simply likes having self-built STK)*, and just not to care enough to set the explicit variable.

It's just a pleasant trivia that improves user and developer experience.

**Comment by Hi-Angel** [comment]
Btw, note that this is port of the according functional in Meson. And Meson nowadays is widely deployed, many core projects such as Mesa, XServer, libinput are using Meson, so you can say such autodetection feature is widely tested as well.

**Comment by qwertychouskie** [comment]
If ccache speeds up builds and doesn't cause bugs, I vote to merge it.

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,12 @@ if(NOT CMAKE_BUILD_TYPE)
     set(CMAKE_BUILD_TYPE "STKRelease")
 endif()
 
+find_program(CCACHE_FOUND "ccache")
+if(CCACHE_FOUND)
+    set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE "ccache")
+    set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK "ccache")
+endif(CCACHE_FOUND)
+
 option(SERVER_ONLY "Create a server only (i.e. no graphics or sound)" OFF)
```
Comment (Chujinda) [inline CMakeLists.txt]: CMakeLists.txt

**Comment by james165527** [comment]
Thanks for the CODES dude !

*Yours sincerely,*
Mark James Agripo.

**Comment by Hi-Angel** [comment]
Thanks for reviews everyone! Note, I don't have write access to the repository, could please someone press the merge button for me?

**Comment by Benau** [comment]
With your way user with ccache installed can never disable it.

What about you just update install.md saying how to export ccache like deve said so everyone is happy?

**Comment by Hi-Angel** [comment]
> With your way user with ccache installed can never disable it.

I can add an option to disable it, something like --no-ccache, would that be okay with you?
 
> What about you just update install.md saying how to export ccache like deve said so everyone is happy?

As I mentioned in [this comment](https://github.com/supertuxkart/stk-code/pull/3885#issuecomment-493255377) it's just pleasant trivia for developers and users who build the game. Presence of ccache is usually transparent, a should not affect build output, or anything, except the speed of rebuilds. But if you think there can be corner-cases, where someone does not want ccache, I can add an option to disable it.

**Comment by Benau** [comment]
Because ccache is a you-know-what-you-are-doing thing.....

**Comment by Hi-Angel** [comment]
> Because ccache is a you-know-what-you-are-doing thing.....

It's really not. As I mentioned at the top post, meson-build based systems all use such autodetection simply because that's how meson works. And right now there's quite a lot of them, because many projects migrated from older autotools to meson. Including core ones, like Mesa and xserver for example.

**Comment by Hi-Angel** [comment]
@Benau for reference, here for example Mesa explicitly enable it in CI https://gitlab.freedesktop.org/mesa/mesa/merge_requests/240

**Comment by dumaosen** [comment]
Just merge it at first. I think U can make it optional later.
I dont think it's good to make a good pr unmerged in a long time because it may cause conflict.

**Comment by Hi-Angel** [comment]
FWIW, I had a similar code [merged to Konsole](https://invent.kde.org/kde/konsole/blob/ef13a09e0178317e40076905ef9cef6efbc99598/CMakeLists.txt#L20), except it's a bit more advanced: in particular there's an option to disable it. If there's an interest, I can update this MR with that version of code.

The relevant discussion https://invent.kde.org/kde/konsole/merge_requests/26

**Pull request closed without merging**